{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pandas import DataFrame, read_csv\n",
    "from urllib.parse import urlencode\n",
    "from jsonapi_client import Session, Filter\n",
    "import html\n",
    "import requests\n",
    "import json"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "MGNIFY_API_BASE = 'https://www.ebi.ac.uk/metagenomics/api/latest/'\n",
    "ena_url = 'https://www.ebi.ac.uk/ena/portal/api/filereport'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_metadata(metadata, key):\n",
    "    for m in metadata:\n",
    "        if m['key'].lower() == key.lower():\n",
    "            value = m['value']\n",
    "            unit = html.unescape(m['unit']) if m['unit'] else \"\"\n",
    "            return \"{value} {unit}\".format(value=value, unit=unit)\n",
    "    return None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ERS3341288: MGYA00594700;MGYA00594735;MGYA00594774: https://www.ebi.ac.uk/metagenomics/api/v1/analyses/MGYA00594700/file/ERR3256289_MERGED_FASTQ_SSU_OTU_TABLE_JSON.biom;https://www.ebi.ac.uk/metagenomics/api/v1/analyses/MGYA00594735/file/ERR3255920_MERGED_FASTQ_SSU_OTU_TABLE_JSON.biom;https://www.ebi.ac.uk/metagenomics/api/v1/analyses/MGYA00594735/file/ERR3255920_MERGED_FASTQ_ITS_ITSoneDB_OTU_TABLE_JSON.biom;https://www.ebi.ac.uk/metagenomics/api/v1/analyses/MGYA00594735/file/ERR3255920_MERGED_FASTQ_ITS_UNITE_OTU_TABLE_JSON.biom;https://www.ebi.ac.uk/metagenomics/api/v1/analyses/MGYA00594774/file/ERR3255616_MERGED_FASTQ_SSU_OTU_TABLE_JSON.biom\n",
      "ERS3341289: MGYA00594699;MGYA00594734;MGYA00594773: https://www.ebi.ac.uk/metagenomics/api/v1/analyses/MGYA00594699/file/ERR3256290_MERGED_FASTQ_SSU_OTU_TABLE_JSON.biom;https://www.ebi.ac.uk/metagenomics/api/v1/analyses/MGYA00594734/file/ERR3255921_MERGED_FASTQ_SSU_OTU_TABLE_JSON.biom;https://www.ebi.ac.uk/metagenomics/api/v1/analyses/MGYA00594734/file/ERR3255921_MERGED_FASTQ_LSU_OTU_TABLE_JSON.biom;https://www.ebi.ac.uk/metagenomics/api/v1/analyses/MGYA00594734/file/ERR3255921_MERGED_FASTQ_ITS_ITSoneDB_OTU_TABLE_JSON.biom;https://www.ebi.ac.uk/metagenomics/api/v1/analyses/MGYA00594734/file/ERR3255921_MERGED_FASTQ_ITS_UNITE_OTU_TABLE_JSON.biom;https://www.ebi.ac.uk/metagenomics/api/v1/analyses/MGYA00594773/file/ERR3255617_MERGED_FASTQ_SSU_OTU_TABLE_JSON.biom\n",
      "ERS3341290: MGYA00594698;MGYA00594733;MGYA00594772: https://www.ebi.ac.uk/metagenomics/api/v1/analyses/MGYA00594698/file/ERR3256291_MERGED_FASTQ_SSU_OTU_TABLE_JSON.biom;https://www.ebi.ac.uk/metagenomics/api/v1/analyses/MGYA00594733/file/ERR3255922_MERGED_FASTQ_SSU_OTU_TABLE_JSON.biom;https://www.ebi.ac.uk/metagenomics/api/v1/analyses/MGYA00594733/file/ERR3255922_MERGED_FASTQ_ITS_ITSoneDB_OTU_TABLE_JSON.biom;https://www.ebi.ac.uk/metagenomics/api/v1/analyses/MGYA00594733/file/ERR3255922_MERGED_FASTQ_ITS_UNITE_OTU_TABLE_JSON.biom;https://www.ebi.ac.uk/metagenomics/api/v1/analyses/MGYA00594772/file/ERR3255618_MERGED_FASTQ_SSU_OTU_TABLE_JSON.biom\n",
      "ERS3341291: MGYA00594732;MGYA00594771: https://www.ebi.ac.uk/metagenomics/api/v1/analyses/MGYA00594732/file/ERR3255923_MERGED_FASTQ_ITS_ITSoneDB_OTU_TABLE_JSON.biom;https://www.ebi.ac.uk/metagenomics/api/v1/analyses/MGYA00594732/file/ERR3255923_MERGED_FASTQ_ITS_UNITE_OTU_TABLE_JSON.biom;https://www.ebi.ac.uk/metagenomics/api/v1/analyses/MGYA00594771/file/ERR3255619_MERGED_FASTQ_SSU_OTU_TABLE_JSON.biom\n",
      "ERS3341292: MGYA00594697;MGYA00594731;MGYA00594770: https://www.ebi.ac.uk/metagenomics/api/v1/analyses/MGYA00594697/file/ERR3256293_MERGED_FASTQ_SSU_OTU_TABLE_JSON.biom;https://www.ebi.ac.uk/metagenomics/api/v1/analyses/MGYA00594731/file/ERR3255924_MERGED_FASTQ_ITS_ITSoneDB_OTU_TABLE_JSON.biom;https://www.ebi.ac.uk/metagenomics/api/v1/analyses/MGYA00594731/file/ERR3255924_MERGED_FASTQ_ITS_UNITE_OTU_TABLE_JSON.biom;https://www.ebi.ac.uk/metagenomics/api/v1/analyses/MGYA00594770/file/ERR3255620_MERGED_FASTQ_SSU_OTU_TABLE_JSON.biom;https://www.ebi.ac.uk/metagenomics/api/v1/analyses/MGYA00594770/file/ERR3255620_MERGED_FASTQ_LSU_OTU_TABLE_JSON.biom\n",
      "ERS3341293: MGYA00594696;MGYA00594730;MGYA00594769: https://www.ebi.ac.uk/metagenomics/api/v1/analyses/MGYA00594696/file/ERR3256294_MERGED_FASTQ_SSU_OTU_TABLE_JSON.biom;https://www.ebi.ac.uk/metagenomics/api/v1/analyses/MGYA00594730/file/ERR3255925_MERGED_FASTQ_SSU_OTU_TABLE_JSON.biom;https://www.ebi.ac.uk/metagenomics/api/v1/analyses/MGYA00594730/file/ERR3255925_MERGED_FASTQ_ITS_ITSoneDB_OTU_TABLE_JSON.biom;https://www.ebi.ac.uk/metagenomics/api/v1/analyses/MGYA00594730/file/ERR3255925_MERGED_FASTQ_ITS_UNITE_OTU_TABLE_JSON.biom;https://www.ebi.ac.uk/metagenomics/api/v1/analyses/MGYA00594769/file/ERR3255621_MERGED_FASTQ_SSU_OTU_TABLE_JSON.biom;https://www.ebi.ac.uk/metagenomics/api/v1/analyses/MGYA00594769/file/ERR3255621_MERGED_FASTQ_LSU_OTU_TABLE_JSON.biom\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample name</th>\n",
       "      <th>lineage</th>\n",
       "      <th>biome</th>\n",
       "      <th>feature</th>\n",
       "      <th>material</th>\n",
       "      <th>package</th>\n",
       "      <th>latitude</th>\n",
       "      <th>longitude</th>\n",
       "      <th>depth</th>\n",
       "      <th>elevation</th>\n",
       "      <th>region</th>\n",
       "      <th>instrument model</th>\n",
       "      <th>investigation type</th>\n",
       "      <th>sequencing method</th>\n",
       "      <th>collection-date</th>\n",
       "      <th>forward reads</th>\n",
       "      <th>reverse reads</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample accession</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ERS3341288</th>\n",
       "      <td>SK-16S-01_S21</td>\n",
       "      <td>root:Environmental:Terrestrial:Soil</td>\n",
       "      <td>arid</td>\n",
       "      <td>agricultural</td>\n",
       "      <td>soil</td>\n",
       "      <td>soil</td>\n",
       "      <td>-26.73</td>\n",
       "      <td>25.99</td>\n",
       "      <td>None</td>\n",
       "      <td>1479.0</td>\n",
       "      <td>South Africa</td>\n",
       "      <td>None</td>\n",
       "      <td>metagenome</td>\n",
       "      <td>Illumina Miseq 16S</td>\n",
       "      <td>2017-02-01</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/fastq/ERR325/006/ERR325...</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/fastq/ERR325/006/ERR325...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERS3341289</th>\n",
       "      <td>SK-16S-02_S22</td>\n",
       "      <td>root:Environmental:Terrestrial:Soil</td>\n",
       "      <td>arid</td>\n",
       "      <td>agricultural</td>\n",
       "      <td>soil</td>\n",
       "      <td>soil</td>\n",
       "      <td>-26.73</td>\n",
       "      <td>25.99</td>\n",
       "      <td>None</td>\n",
       "      <td>1479.0</td>\n",
       "      <td>South Africa</td>\n",
       "      <td>None</td>\n",
       "      <td>metagenome</td>\n",
       "      <td>Illumina Miseq 16S</td>\n",
       "      <td>2017-02-01</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/fastq/ERR325/007/ERR325...</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/fastq/ERR325/007/ERR325...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERS3341290</th>\n",
       "      <td>SK-16S-03_S23</td>\n",
       "      <td>root:Environmental:Terrestrial:Soil</td>\n",
       "      <td>arid</td>\n",
       "      <td>agricultural</td>\n",
       "      <td>soil</td>\n",
       "      <td>soil</td>\n",
       "      <td>-26.73</td>\n",
       "      <td>25.99</td>\n",
       "      <td>None</td>\n",
       "      <td>1479.0</td>\n",
       "      <td>South Africa</td>\n",
       "      <td>None</td>\n",
       "      <td>metagenome</td>\n",
       "      <td>Illumina Miseq 16S</td>\n",
       "      <td>2017-02-01</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/fastq/ERR325/008/ERR325...</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/fastq/ERR325/008/ERR325...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERS3341291</th>\n",
       "      <td>SK-16S-04_S24</td>\n",
       "      <td>root:Environmental:Terrestrial:Soil</td>\n",
       "      <td>arid</td>\n",
       "      <td>agricultural</td>\n",
       "      <td>soil</td>\n",
       "      <td>soil</td>\n",
       "      <td>-26.73</td>\n",
       "      <td>25.99</td>\n",
       "      <td>None</td>\n",
       "      <td>1479.0</td>\n",
       "      <td>South Africa</td>\n",
       "      <td>None</td>\n",
       "      <td>metagenome</td>\n",
       "      <td>Illumina Miseq 16S</td>\n",
       "      <td>2017-02-01</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/fastq/ERR325/009/ERR325...</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/fastq/ERR325/009/ERR325...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERS3341292</th>\n",
       "      <td>SK-16S-05_S25</td>\n",
       "      <td>root:Environmental:Terrestrial:Soil</td>\n",
       "      <td>arid</td>\n",
       "      <td>agricultural</td>\n",
       "      <td>soil</td>\n",
       "      <td>soil</td>\n",
       "      <td>-26.73</td>\n",
       "      <td>25.99</td>\n",
       "      <td>None</td>\n",
       "      <td>1479.0</td>\n",
       "      <td>South Africa</td>\n",
       "      <td>None</td>\n",
       "      <td>metagenome</td>\n",
       "      <td>Illumina Miseq 16S</td>\n",
       "      <td>2017-02-01</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/fastq/ERR325/000/ERR325...</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/fastq/ERR325/000/ERR325...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERS3341293</th>\n",
       "      <td>SK-16S-06_S26</td>\n",
       "      <td>root:Environmental:Terrestrial:Soil</td>\n",
       "      <td>arid</td>\n",
       "      <td>agricultural</td>\n",
       "      <td>soil</td>\n",
       "      <td>soil</td>\n",
       "      <td>-26.73</td>\n",
       "      <td>25.99</td>\n",
       "      <td>None</td>\n",
       "      <td>1479.0</td>\n",
       "      <td>South Africa</td>\n",
       "      <td>None</td>\n",
       "      <td>metagenome</td>\n",
       "      <td>Illumina Miseq 16S</td>\n",
       "      <td>2017-02-01</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/fastq/ERR325/001/ERR325...</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/fastq/ERR325/001/ERR325...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    sample name                              lineage  biome  \\\n",
       "sample accession                                                              \n",
       "ERS3341288        SK-16S-01_S21  root:Environmental:Terrestrial:Soil  arid    \n",
       "ERS3341289        SK-16S-02_S22  root:Environmental:Terrestrial:Soil  arid    \n",
       "ERS3341290        SK-16S-03_S23  root:Environmental:Terrestrial:Soil  arid    \n",
       "ERS3341291        SK-16S-04_S24  root:Environmental:Terrestrial:Soil  arid    \n",
       "ERS3341292        SK-16S-05_S25  root:Environmental:Terrestrial:Soil  arid    \n",
       "ERS3341293        SK-16S-06_S26  root:Environmental:Terrestrial:Soil  arid    \n",
       "\n",
       "                        feature material package latitude longitude depth  \\\n",
       "sample accession                                                            \n",
       "ERS3341288        agricultural     soil    soil   -26.73     25.99   None   \n",
       "ERS3341289        agricultural     soil    soil   -26.73     25.99   None   \n",
       "ERS3341290        agricultural     soil    soil   -26.73     25.99   None   \n",
       "ERS3341291        agricultural     soil    soil   -26.73     25.99   None   \n",
       "ERS3341292        agricultural     soil    soil   -26.73     25.99   None   \n",
       "ERS3341293        agricultural     soil    soil   -26.73     25.99   None   \n",
       "\n",
       "                 elevation         region instrument model investigation type  \\\n",
       "sample accession                                                                \n",
       "ERS3341288         1479.0   South Africa              None        metagenome    \n",
       "ERS3341289         1479.0   South Africa              None        metagenome    \n",
       "ERS3341290         1479.0   South Africa              None        metagenome    \n",
       "ERS3341291         1479.0   South Africa              None        metagenome    \n",
       "ERS3341292         1479.0   South Africa              None        metagenome    \n",
       "ERS3341293         1479.0   South Africa              None        metagenome    \n",
       "\n",
       "                    sequencing method collection-date  \\\n",
       "sample accession                                        \n",
       "ERS3341288        Illumina Miseq 16S      2017-02-01    \n",
       "ERS3341289        Illumina Miseq 16S      2017-02-01    \n",
       "ERS3341290        Illumina Miseq 16S      2017-02-01    \n",
       "ERS3341291        Illumina Miseq 16S      2017-02-01    \n",
       "ERS3341292        Illumina Miseq 16S      2017-02-01    \n",
       "ERS3341293        Illumina Miseq 16S      2017-02-01    \n",
       "\n",
       "                                                      forward reads  \\\n",
       "sample accession                                                      \n",
       "ERS3341288        ftp.sra.ebi.ac.uk/vol1/fastq/ERR325/006/ERR325...   \n",
       "ERS3341289        ftp.sra.ebi.ac.uk/vol1/fastq/ERR325/007/ERR325...   \n",
       "ERS3341290        ftp.sra.ebi.ac.uk/vol1/fastq/ERR325/008/ERR325...   \n",
       "ERS3341291        ftp.sra.ebi.ac.uk/vol1/fastq/ERR325/009/ERR325...   \n",
       "ERS3341292        ftp.sra.ebi.ac.uk/vol1/fastq/ERR325/000/ERR325...   \n",
       "ERS3341293        ftp.sra.ebi.ac.uk/vol1/fastq/ERR325/001/ERR325...   \n",
       "\n",
       "                                                      reverse reads  \n",
       "sample accession                                                     \n",
       "ERS3341288        ftp.sra.ebi.ac.uk/vol1/fastq/ERR325/006/ERR325...  \n",
       "ERS3341289        ftp.sra.ebi.ac.uk/vol1/fastq/ERR325/007/ERR325...  \n",
       "ERS3341290        ftp.sra.ebi.ac.uk/vol1/fastq/ERR325/008/ERR325...  \n",
       "ERS3341291        ftp.sra.ebi.ac.uk/vol1/fastq/ERR325/009/ERR325...  \n",
       "ERS3341292        ftp.sra.ebi.ac.uk/vol1/fastq/ERR325/000/ERR325...  \n",
       "ERS3341293        ftp.sra.ebi.ac.uk/vol1/fastq/ERR325/001/ERR325...  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = DataFrame(columns=('sample name',\n",
    "                        'lineage',\n",
    "                        'biome',\n",
    "                        'feature',\n",
    "                        'material',\n",
    "                        'package',\n",
    "                        'latitude',\n",
    "                        'longitude',\n",
    "                        'depth',\n",
    "                        'elevation',\n",
    "                        'region',\n",
    "                        'instrument model',\n",
    "                        'investigation type',\n",
    "                        'sequencing method',\n",
    "                        'collection-date',\n",
    "                        'forward reads',\n",
    "                        'reverse reads'))\n",
    "df.index.name = 'sample accession'\n",
    "\n",
    "depth_label = 'geographic location (depth)'\n",
    "latitude_label = 'geographic location (latitude)'\n",
    "longitude_label = 'geographic location (longitude)'\n",
    "region_label = 'geographic location (country and/or sea,region)'\n",
    "date_label = 'collection date'\n",
    "elevation_label = 'elevation'\n",
    "env_biome_label = 'environment (biome)'\n",
    "env_feature_label = 'environment (feature)'\n",
    "env_material_label = 'environment (material)'\n",
    "env_package_label = 'environmental package'\n",
    "instrument_model_label = 'instrument model'\n",
    "investigation_label = 'investigation type'\n",
    "sequence_method_label = 'sequencing method'\n",
    "\n",
    "with Session(MGNIFY_API_BASE) as mg:\n",
    "    params = {\n",
    "        'metadata_key': 'investigation type',\n",
    "        'metadata_value': 'metagenome',\n",
    "        'lineage': 'root:Environmental:Terrestrial',\n",
    "        'include': 'runs'\n",
    "    }\n",
    "    fltr = Filter(urlencode(params))\n",
    "    \n",
    "    i = 0\n",
    "    for sample in mg.iterate('samples', fltr):\n",
    "            if i > 5: break\n",
    "            i += 1\n",
    "        #try:\n",
    "            ena_params = {\n",
    "                'accession': sample.accession,\n",
    "                'result': 'read_run',\n",
    "                'fields': 'fastq_ftp',\n",
    "                'format': 'JSON'\n",
    "            }\n",
    "            \n",
    "            data = json.loads(requests.get(ena_url, params=ena_params).text)\n",
    "            fastq_links = ';'.join([d['fastq_ftp'] for d in data])\n",
    "            fastq_links = fastq_links.split(';')\n",
    "            forward_reads = ';'.join(list(filter(lambda s: s[-11:-9] != '_2', fastq_links)))\n",
    "            reverse_reads = ';'.join(list(filter(lambda s: s[-11:-9] == '_2', fastq_links)))\n",
    "            \n",
    "            #runs_link = sample.relationships.runs.links.related\n",
    "            analyzes = []\n",
    "            downloads = []\n",
    "            print(sample.accession, end=': ')\n",
    "            for run in mg.iterate(f'samples/{sample.accession}/runs'):\n",
    "                for an in mg.iterate(f'runs/{run.id}/analyses'):\n",
    "                    analyzes.append(an.id)\n",
    "                    for dnl in mg.iterate(f'analyses/{an.id}/downloads'):\n",
    "                        #print(dnl.links.self, end=';')\n",
    "                        if dnl.attributes.file_format.name == 'JSON Biom':\n",
    "                            downloads.append(str(dnl.links.self))\n",
    "            analyzes = ';'.join(analyzes)\n",
    "            downloads = ';'.join(downloads)\n",
    "            print(analyzes, downloads, sep=': ')\n",
    "            \n",
    "            df.loc[sample.accession] = [\n",
    "                sample.sample_name,\n",
    "                sample.biome.id,\n",
    "                get_metadata(sample.sample_metadata, env_biome_label),\n",
    "                get_metadata(sample.sample_metadata, env_feature_label),\n",
    "                get_metadata(sample.sample_metadata, env_material_label),\n",
    "                get_metadata(sample.sample_metadata, env_package_label),\n",
    "                get_metadata(sample.sample_metadata, latitude_label),\n",
    "                get_metadata(sample.sample_metadata, longitude_label),\n",
    "                get_metadata(sample.sample_metadata, depth_label),\n",
    "                get_metadata(sample.sample_metadata, elevation_label),\n",
    "                get_metadata(sample.sample_metadata, region_label),\n",
    "                get_metadata(sample.sample_metadata, instrument_model_label),\n",
    "                get_metadata(sample.sample_metadata, investigation_label),\n",
    "                get_metadata(sample.sample_metadata, sequence_method_label),\n",
    "                get_metadata(sample.sample_metadata, date_label),\n",
    "                forward_reads,\n",
    "                reverse_reads\n",
    "            ]\n",
    "        #except Exception:\n",
    "           # print(sample.accession)\n",
    "            #continue\n",
    "\n",
    "df"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
